### DISCUSSION TITLE
How to implement Caching?

### DISCUSSION THREAD

**Discussion by dinhkhanh** [payloadcms/payload#1436] [General]
I'm considering PayloadCMS for my next projects.
Reading through the documents, with some research, I can't find anything related to Caching.

So, is there any built-in caching in Payload? Or, is there any plugin for caching? Or, how to create a plugin, or implement caching for Payload?

By caching, I mean DB queries cache. And I'd like to use Redis.

**Comment by iakshay** [comment]
Perhaps something like this could work - https://dev.to/ahmedmagdy11/building-cache-layer-using-redis-and-mongoose-11kb

where you override the `exec` function in Mongoose. Albeit, this is hacky but should be do-able.

**Comment by sangallimarco** [comment]
I suggest to cache the responses from the PayloadCMS internal API.

So here's what we did:

- An express middleware that return cached responses from REDIS (It has to be added before  PayloadCMS routes) 
- Custom express controllers that store the response from the payload internal API
- Drop the REDIS cache using a custom collection hook (afterChange)

**Reply by dinhkhanh** [reply]
Do you have any plan to release it as a package?

**Reply by sangallimarco** [reply]
I guess I can  create a PR soon and point you at it.
Have you got any special requests? Can you give me a brief description of the scenario?

**Reply by dinhkhanh** [reply]
It's just a common use case, nothing any special. 

The package should be able to:

- Configure the redis host and some other preferences / options.
- Allow to purge all caches.
- Allow to invalidate item / collection cache after updating.

**Reply by sangallimarco** [reply]
Got it, thanks. I'll write an update shortly

**Reply by sangallimarco** [reply]
We have just published a new package on Npm
https://github.com/Aengz/payload-redis-cache

**Reply by dinhkhanh** [reply]
Thank you. I'll check it out now.

**Reply by sangallimarco** [reply]
A new version with excludedCollections has just been released.

**Reply by gusfune** [reply]
Is it still being maintained @sangallimarco ? See it's been a while without updates.

**Reply by TechSynthesis** [reply]
Hey @sangallimarco any updates planned for v3 ?

**Comment by ssyberg** [comment]
This is old but I'd like to bump it! We run a fairly high traffic site and our AWS costs are about 5x higher than we expected when we first started down the Payload path.

**Reply by cronin4392** [reply]
@ssyberg I was doing some digging and stumbled on a [plugin someone made for redis caching](https://github.com/Aengz/payload-redis-cache).

**Reply by thijssmudde** [reply]
This does not work for v3

**Comment by DannyZB** [comment]
Did any of you guys try caching on Nginx?
With an identical request, you can create a regular nginx proxy cache, depending on whether you repeatedly use similar filters.

**Comment by willemmulder** [comment]
For v3, there is now a discussion here: https://github.com/payloadcms/payload/discussions/9909.
